Add event testing

weigandjakob opened the pull request.

egekorkan commented on src/Tester.ts:
```diff
@@ -10,75 +10,369 @@ After the test a test report can be generated and analyzed to get more meaning o
+            // Checking if data package validates against its schema.
+            var validationError: Array<any>
+            if (testMode == Utils.InteractionType.Event)
+                validationError = Utils.validateResponse(interactionName, receivedData, self.testConfig.SchemaLocation, Utils.SchemaType.EventData)
```
Only realized here. If there are two interactions with the same name, this stuff fails. This is generally not the case so it an assumption can be made

weigandjakob replied:
I do not think so. The interaction is identified by a tuple of schemaType and name. Per schema type only one name is possible. This is a side effect of my changes to extractSchema(), which were done to be able to be able to support interactions with more than one schema (event -> TD example 35) I did not really think about it but if a property could also have a schema for subscription and cancellation I would suggest adding a propertySubscription and propertyCancellation schemaType in the future. I noticed your described behavior and changed it for getInteractionByName() which no uses a tuple of interactionType and name.

egekorkan replied:
I see, that sounds good. For properties, there is no plan to add such information

egekorkan commented:
Running it also works fine. Could you also add a faulty thing for the events?

weigandjakob commented on src/Tester.ts:
```diff
@@ -10,75 +10,369 @@ After the test a test report can be generated and analyzed to get more meaning o
+     * Logs a message in blue.
+     * @param message The message to log.
+     */
+    private log(message: string): void {
```
Thought the same but two downsides: 
1. The smaller one: every log statement is longer again
2. I thought log() could potentially do more in the future like also handling a prefix, I made this change in the latest commit.

As always I can still change the method name if you want.

egekorkan replied:
ok let's leave it like this

weigandjakob commented on src/wot-test-bench.ts:
```diff
@@ -43,44 +43,43 @@ srv.start().then((WoT) => {
                 type: "string",
-                writable: false,
-                description: "(not finished) Shows the status of the test" + "bench whether it is currently testing a device or not",
+                writeOnly: false,
```
You meant readOnly: true did you? Your comment is between writeOnly and readOnly :D.

weigandjakob replied:
I made my expected change in the newest commit, just wanted to ask to make sure.

egekorkan replied:
yes I meant readOnly :)

weigandjakob commented on src/wot-test-bench.ts:
```diff
@@ -43,44 +43,43 @@ srv.start().then((WoT) => {
-                writable: false,
-                description: "Contains all of the outputs of the testing." + "Not necessary for fastTest",
+                writeOnly: false,
+                readOnly: false,
```
Resolved by newest commit.

weigandjakob commented on src/wot-test-bench.ts:
```diff
@@ -91,17 +90,16 @@ srv.start().then((WoT) => {
                 },
-                description:
-                    "By invoking this action, the test bench consumes" + "the thing under test, generates data to be" + "sent. Not necessary for fastTest",
+                description: "By invoking this action, the test bench consumes the thing under test, generates data to be sent. Not necessary for fastTest",
```
Resolved by newest commit.

weigandjakob commented on src/wot-test-bench.ts:
```diff
@@ -110,7 +108,7 @@ srv.start().then((WoT) => {
                     type: "boolean",
                 },
-                description: "By invoking this action the testing starts" + "and produces a test report that can be read." + "Not necessary for fastTest",
+                description: "By invoking this action the testing starts and produces a test report that can be read. Not necessary for fastTest",
```
Resolved by newest commit.

weigandjakob commented on src/TestReport.ts:
```diff
@@ -35,11 +28,42 @@ export class Payload {
+/**
+ * A mini test report only containing a passed boolean, two payload objects, containing the sent and received payload and a testResult object.
+ */
 export class MiniTestReport {
```
In the newest commit I removed micro test report and added a getPrintableOnlyOut() function to MiniTestReport. I think now it should be better?

egekorkan replied:
yes!

weigandjakob commented:
@egekorkan faulty Thing was still on the old node-wot version. I did convert it to the new version but I am not sure which of the other bugs should be fixed and which are considered intentional faults....

I added a new event to faulty Thing that leads the testbench to fail that test but due to the other bugs/intentional faults in faultyThing (see my remark above) I can not finish a testrun with faultyThing.  For perfectThing I also added a new event so that it has now two passing event test one with a timeout during subscription and one with a succesful subscription and received data. This can be changed if the timeout is not in the "spirit" of perfectThing...

weigandjakob commented:
Small additional remark: I noticed that Travis CI now has the new config with the prettier write command (- npx prettier --write .). I do not know if that command does anything at this stage (does it commit potential changes made in this command?) or if it should be removed.

egekorkan commented:
perfectThing does not need to change but faultyThing needs to be adapted to the newest version, otherwise we cannot show how a report should look like for beginners

egekorkan commented:
Regarding prettier, could you do test commit that does not obey the style?

weigandjakob commented:
So the philosophy of faultyThing is that it works but every single test fails?

weigandjakob commented:
I added the not style obeying commit. Travis CI ran the command but did not commit any changes.

egekorkan commented:
> @egekorkan So the philosophy of faultyThing is that it works but every single test fails?

So not all tests necessarily but would be nice to show different tests that fail, i.e. property with wrong data, property which has different written value once read, action with wrong  output, action with timeout, event wrong data, (maybe events not emitting)

weigandjakob commented on src/Tester.ts:
```diff
@@ -257,15 +257,19 @@ export class Tester {
+                )
                 container.passed = false
                 container.eventDataReport.passed = false
                 container.eventDataReport.received.push(new EventData(receivedTimeStamp, JSON.parse(JSON.stringify(receivedData, null, " ")), result))
```
@egekorkan
I changed this if block after noticing some inaccuracies. Could you please give feedback on this?

On this topic: I checked the old behavior before any of my changes for testAction(). There is a check if the returned value is JSON. If not, there is still a message added which has the returned non JSON value as received value. I removed this due to concerns this could be a problem when the returned value is then added to the JSON testReport. 

The reason for my question is this if block (`if (testMode == Utils.InteractionType.Event && !interaction.hasOwnProperty("data"))`). Theoretically the received data could be added to the container here but this would be before the check if it is a JSON object. If I put it behind the JSON check it could be that the result just says "no JSON format" but not that no data was expected. Is it possible to receive non JSON conformant data here? (undefined is rendered as null...) Alternatively I could keep the slightly inelegant JSON.parse(JSON.stringify(...)) solution.

Due to action output being optional in a TD i would also add this part to testAction() when you gave feedback...

weigandjakob replied:
I noticed every information on a property is optional (due to PropertyAffordance being a subclass to DataSchema TD Spec §5.3.1.3). This would mean there also could be an empty property? Is this the case and if yes, I do not think the testbench can handle this. For an example the Thing Description Playground validates this TD: 
```
{
    "title": "TestServient",
    "description": "Test servient that can be used as a servient to be tested with the WoT Test Bench",
    "properties": {
        "display": {
            "observable": false,
            "readOnly": false,
            "writeOnly": false,
            "forms": [
                {
                    "href": "http://localhost:8083/TestServient/properties/display",
                    "contentType": "application/json",
                    "op": ["readproperty", "writeproperty"]
                }
            ]
        }
    },
    "id": "urn:uuid:560290c8-6490-4a58-99ca-eea9bc8c25d2",
    "@context": "https://www.w3.org/2019/wot/td/v1",
    "@type": "Thing",
    "security": ["nosec_sc"],
    "forms": [
        {
            "href": "http://localhost:8083/TestServient/all/properties",
            "contentType": "application/json",
            "op": ["readallproperties", "readmultipleproperties", "writeallproperties", "writemultipleproperties"]
        }
    ],
    "securityDefinitions": {
        "nosec_sc": {
            "scheme": "nosec"
        }
    }
}

```
![image](https://user-images.githubusercontent.com/65398308/91364043-2fed1f00-e7fe-11ea-8660-568fb7f4faec.png)

egekorkan replied:
regarding the first comment, I do not understand the question but:
>Alternatively I could keep the slightly inelegant JSON.parse(JSON.stringify(...)) solution if this solution is guaranteed to work with every possible input.

There is no such guarantee. The received data as a buffer gets converted as JSON by content serdes so that the script can use it as a JSON. We can assume that we always get JSON but this should be clearly stated.

weigandjakob commented on testing-files/faultyThing.js:
```diff
@@ -201,23 +112,97 @@ srv.start()
+            thing.setPropertyWriteHandler("wrongWritable", async () => {
+                console.log("Writing the old value.")
+                return 15
+            })
```
@egekorkan Is this the correct use of a propertyHandler? I could observe two strange behaviors with this function: 

1. when trying to call `thing.writeProperty("wrongWritable"´, 15)` in the handler function the Thing logs "Writing the old value" about 40 times and then crashes when this handler is called. 

2. when nothing is returned e. g. `return` instead of `return 15` (without a call of `thing.writeProperty(...)`) the testbench receives `undefined` for second read and thus the test result is "wrong type" instead of "second read did not match write". This is strange because in this writeHandler nothing happend the original value should be unchanged.

Is this potentially another issue in node-wot?

egekorkan replied:
The correct way should be with resolve(). The value in resolve is the actual value written.

weigandjakob replied:
Ah ok wenn writing `return` (without value) in an async function or just `resolve()` (without value) in a promise node-wot probably writes null thus the output was correct.

egekorkan commented:
> This would mean there also could be an empty property?

This would mean that the value gotten should be either a JSON null or there should be no value at all, like an action not having any input. It is a rare case that generally does not happen. Could be logged as a warning message.

egekorkan commented:
> @egekorkan I added the not style obeying commit. Travis CI ran the command but did not commit any changes.

In this case, I would still leave the prettifier command.

weigandjakob commented on README.md:
```diff
@@ -105,50 +110,119 @@ ___
-- You can use your browser and the GET requests to inspect all properties during the procedure.
+### TLDR
 
+<img src="readme-images/test.svg" align="center" height="400" width="400">
```
@egekorkan I added a readme-images folder on root is this ok, or should this be changed? When using svg for images I noticed, I can only give size information if using html instead of markdown directly...

egekorkan replied:
That is fine :+1:

weigandjakob commented:
> In this case, I would still leave the prettifier command.

Ok.

weigandjakob commented on README.md:
```diff
@@ -105,50 +110,119 @@ ___
+### TLDR
 
+<img src="readme-images/testingProcess.svg" align="center" height="400" width="400">
 
```
@egekorkan Is this sufficient for an explanation how events are tested? I also explained actions and read/writeProperty to have a complete overview...

weigandjakob replied:
I meant the whole "How does the testing work?" chapter, not only the image line.

egekorkan replied:
I would like to see a more thorough illustration of how events are tested, it can be reused in other places that way ;)

weigandjakob replied:
Do you mean I should add another image exclusively for event testing or describe the process in more detail in the text or both?

weigandjakob replied:
I added an eventTesting image to the README was this the kind of illustration you had in mind?

egekorkan replied:
It is fine for now, let's iterate on it later.

weigandjakob replied:
Ok.

weigandjakob commented on src/Tester.ts:
```diff
@@ -10,75 +10,410 @@ After the test a test report can be generated and analyzed to get more meaning o
+                            result
+                        )
+                    )
+                    return
```
@egekorkan this refers to the question where you wrote you did not unerstand it. I need feedback concerning this section (from `if (testMode == Utils.InteractionType.Event && !interaction.hasOwnProperty("data")) {` up to here. I added an extra JSON check to make sure I do send a non JSON value...

I will ask pretty much the same question on affected part of testAction().

egekorkan replied:
Ah ok, all good!

weigandjakob commented on src/Tester.ts:
```diff
@@ -88,461 +423,412 @@ export class Tester {
+                    )
+                    container.report.received = receivedData
+                    container.report.result = new Result(99, "Received unexpected (action did not have an output property in TD) return value.")
+                    return
```
@egekorkan this refers to the question where you wrote you did not unerstand it. I need feedback concerning this section from `if (!interaction.hasOwnProperty("output")) {` up to here. I added an extra JSON check to make sure I do send a non JSON value...

I will ask pretty much the same question on affected part of handleReceivedData().

weigandjakob commented:
@egekorkan 
> This would mean that the value gotten should be either a JSON null or there should be no value at all, like an action not having any input. It is a rare case that generally does not happen. Could be logged as a warning message.

It seems like testbench generates a request and response schema of type undefined. This means it should validate against the received payload (being also undefined in this case...). So I think there is no change necessary for properties.

weigandjakob commented on testing-files/faultyThing.js:
```diff
@@ -201,23 +128,98 @@ srv.start()
+        .catch((err) => {
+            throw "Could not connect to servient"
+        })
+})
```
@egekorkan The faultyThing is no ready in my opinion. I added three observable properties one alternates between being set violate minimum and maximum, one emits wrong data type and one emits nothing at all. Additionally I added an event that emits the wrong data Type. Is this sufficient?

egekorkan replied:
Yes, good work!

weigandjakob commented on testing-files/perfectThing.js:
```diff
@@ -168,6 +172,11 @@ srv.start().then((WoT) => {
+
             thing.expose().then(() => {
                 console.info(thing.title + " ready")
             })
```
@egekorkan In my opinion perfectThing is ready. It has two observableProperties and two events one of each times out the other emits valid data. Is this sufficient?

egekorkan replied:
Yes, great work!

egekorkan commented:
Could you also bump version in package.json to 1.1.0?

weigandjakob commented:
> Could you also bump version in package.json to 1.1.0?

Yes new commit has the new version.
